solver/src/naive_solver/multi_order_solver.rs

```diff
@@ -0,0 +1,369 @@
+#![allow(dead_code)]
+
+use super::single_pair_settlement::{AmmSwapExactTokensForTokens, SinglePairSettlement};
+use crate::settlement::Trade;
+use anyhow::{anyhow, Result};
+use model::order::{OrderCreation, OrderKind};
+use num::{bigint::Sign, BigInt};
+use std::collections::HashMap;
+use web3::types::{Address, U256};
+
+#[derive(Default)]
+struct TokenContext {
+    address: Address,
+    reserve: BigInt,
+    buy_volume: BigInt,
+    sell_volume: BigInt,
+}
+
+impl TokenContext {
+    pub fn is_excess(&self, other: &TokenContext) -> bool {
+        &self.reserve * (&other.sell_volume - &other.buy_volume)
+            < &other.reserve * (&self.sell_volume - &self.buy_volume)
+    }
+}
+
+/**
+ * Computes a settlement using orders of a single pair and the direct AMM between those tokens.
+ * Returns an error if computation fails, orders are not already filtered for a specific token pair, or the reserve information for that
+ * pair is not available.
+ */
+pub fn solve(
+    orders: impl Iterator<Item = OrderCreation> + Clone,
+    reserves: &HashMap<Address, U256>,
+) -> Result<SinglePairSettlement> {
+    let (context_a, context_b) = split_into_contexts(orders.clone(), reserves)?;
+    let (shortage, excess) = if context_a.is_excess(&context_b) {
+        (context_b, context_a)
+    } else {
+        (context_a, context_b)
+    };
+    let uniswap_out = compute_uniswap_out(&shortage, &excess);
+    let uniswap_in = bigint_to_u256(&compute_uniswap_in(&uniswap_out, &shortage, &excess))?;
+    let uniswap_out = bigint_to_u256(&uniswap_out)?;
+    let interaction = if uniswap_out > U256::zero() {
```

Comment:
I'm ok with either version just pointing out this alternative
```suggestion
    let interaction = if !uniswap_out.is_zero() {
```